archive/100-theorems-list/83_friendship_graphs.lean

```diff
@@ -76,8 +76,10 @@ theorem adj_matrix_sq_of_ne {v w : V} (hvw : v ≠ w) :
   ((G.adj_matrix R) ^ 2) v w = 1 :=
 begin
   rw [sq, ← nat.cast_one, ← hG hvw],
-  simp [common_neighbors, neighbor_finset_eq_filter, finset.filter_filter, finset.filter_inter,
-    and_comm, ← neighbor_finset_def],
+  simp only [mul_eq_mul, mul_adj_matrix_apply, adj_matrix_apply, sum_boole, fintype.card_of_finset],
+  congr' 2,
+  ext u,
+  simp [common_neighbors, and_comm],
```

Comment:
What happened here? This feels like a regression if it's no longer possible to finish by a `simp`.

Reply:
This ended up revealing a subtle bug in the `fintype` instances that's been affecting us for a while, where we get `s x` instead of `x ∈ s` when simplifying expressions involving certain finite sets. I had added an extra `fintype` instance in this PR (`fintype_inter_of_right`) to mirror a pre-existing one, but it involved a rewrite so inhibited `fintype.card_of_finset` from applying in this case (which I didn't realize when I rewrote the proof like this). However, it turned out that these instances were defined improperly, using the sets directly as predicates. I'm waiting for a local recompile to be able to test whether it fixes `83_friendship_graphs`.

The commit I just pushed was from before I found this bug, and it fixes the above proof by rewriting using `fintype.card_of_finset` first, since  `simp` was sensitive to the simplification order.

Reply:
Ok, this is back to being a single simp!

I'm not sure if all of mathlib compiles yet -- there's a chance there are preexisting workarounds for this bug.

Reply:
Also, it looks like what happened is that when that improvement to congruence lemmas landed, `fintype.card_of_finset` ended up being more likely to apply. When Gabriel bumped up the mathlib version, you can see he added `mem_neighbor_set'` as a workaround for the set API breakage induced by the `fintype` instances. https://github.com/leanprover-community/mathlib/commit/b71077179fac7179e3a342ff319392ad8dcb773c